pkg/controller/eks/cluster.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errGetProviderSecret)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	eksClient, stsClient, err := c.newClientFn(ctx, s.Data[p.Spec.CredentialsSecretRef.Key], p.Spec.Region, awsclients.UseProviderSecret)
+	return &external{client: eksClient, sts: stsClient, kube: c.kube}, errors.Wrap(err, errCreateEKSClient)
+}
+
+type external struct {
+	client eks.Client
+	sts    eks.STSClient
+	kube   client.Client
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.Cluster)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotEKSCluster)
+	}
+
+	rsp, err := e.client.DescribeClusterRequest(&awseks.DescribeClusterInput{Name: aws.String(meta.GetExternalName(cr))}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(eks.IsErrorNotFound, err), errDescribeFailed)
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	eks.LateInitialize(&cr.Spec.ForProvider, rsp.Cluster)
+	if !reflect.DeepEqual(current, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errKubeUpdateFailed)
+		}
+	}
+
+	cr.Status.AtProvider = eks.GenerateObservation(rsp.Cluster)
+	switch cr.Status.AtProvider.Status {
+	case v1beta1.ClusterStatusActive:
+		cr.Status.SetConditions(runtimev1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.ClusterStatusCreating:
+		cr.Status.SetConditions(runtimev1alpha1.Creating())
+	case v1beta1.ClusterStatusDeleting:
+		cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	default:
+		cr.Status.SetConditions(runtimev1alpha1.Unavailable())
+	}
+	upToDate, err := eks.IsUpToDate(&cr.Spec.ForProvider, rsp.Cluster)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errUpToDateFailed)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  upToDate,
+		ConnectionDetails: eks.GetConnectionDetails(rsp.Cluster, e.sts),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1beta1.Cluster)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotEKSCluster)
+	}
+	cr.SetConditions(runtimev1alpha1.Creating())
+	if cr.Status.AtProvider.Status == v1beta1.ClusterStatusCreating {
+		return managed.ExternalCreation{}, nil
+	}
+	_, err := e.client.CreateClusterRequest(eks.GenerateCreateClusterInput(meta.GetExternalName(cr), &cr.Spec.ForProvider)).Send(ctx)
+	return managed.ExternalCreation{}, errors.Wrap(err, errCreateFailed)
```

Comment:
It could be worth skipping `AlreadyExists` error if the `Create` request is not idempotent.

Reply:
@muvaf I think this will be handled by checking if the resource is in `CREATING` state in the line above right? On this topic though, I dropped a comment on another PR about the ability to provide an `idempotencyToken` on some AWS resources. I opted not to use it here because it seems that these requests are already idempotent, and reasoning about when to change the token actually adds a pretty significant layer of complexity that may require further investigation about addressing idempotency across all Crossplane. However, wanted to get your take on it: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2@v0.23.0/service/eks?tab=doc#CreateClusterInput

Reply:
> I think this will be handled by checking if the resource is in CREATING state in the line above right?

Yes, it's fine either way. I personally just skip if I know that in 100% of the cases I'd like to skip that error even though it's unlikely to occur. But it can be either way.

>  I opted not to use it here because it seems that these requests are already idempotent, and reasoning about when to change the token actually adds a pretty significant layer of complexity that may require further investigation about addressing idempotency across all Crossplane. However, wanted to get your take on it

Yeah, I noticed that in the API and I agree with you that it adds some complexity that may not be really necessary. Maybe `resourceVersion` could be used as token there but it's probably just fine as is.